src/bms/unparse.rs

```diff
@@ -0,0 +1,1313 @@
+///
+/// # PROCESSING FLOW OVERVIEW:
+/// 1. **GROUP EVENTS**: Events are grouped by track, channel, and non-strictly increasing time
+/// 2. **SPLIT INTO MESSAGE SEGMENTS**: Each group is further split into message segments with stricter rules:
+///    - Strictly increasing time (prevents overlaps)
+///    - Consistent denominators (ensures accurate representation)
+/// 3. **GENERATE TOKENS**: Each message segment becomes one Token::Message with all events encoded
+///
+/// Arguments:
+///     events: An iterator yielding (&time, &event) pairs to process
+///     id_allocation: Optional tuple containing (token_creator, key_extractor, id_manager) for ID allocation mode
+///     channel_mapper: Function to map events to channels
+///     message_formatter: Function to format events into [char; 2]
+///
+/// Returns:
+///     EventProcessingResult containing message_tokens, late_def_tokens, and updated maps
+///
+/// The function leverages Rust's iterator chains for efficient processing and supports
+/// both ID-based and direct value-based event processing.
+fn build_event_messages<
+    'a,
+    Event: 'a,
+    Key: 'a + ?Sized + std::hash::Hash + Eq,
+    EventIterator,
+    TokenCreator,
+    KeyExtractor,
+    ChannelMapper,
+    MessageFormatter,
+>(
+    event_iter: EventIterator,
+    mut id_allocation: Option<(TokenCreator, KeyExtractor, &mut ObjIdManager<'a, Key>)>,
+    channel_mapper: ChannelMapper,
+    message_formatter: MessageFormatter,
+) -> EventProcessingResult<'a>
+where
+    EventIterator: Iterator<Item = (&'a ObjTime, &'a Event)>,
+    TokenCreator: Fn(ObjId, &'a Key) -> Token<'a>,
+    KeyExtractor: Fn(&'a Event) -> &'a Key,
+    ChannelMapper: Fn(&'a Event) -> Channel,
+    MessageFormatter: Fn(&'a Event, Option<ObjId>) -> [char; 2],
+{
+    let mut late_def_tokens: Vec<Token<'a>> = Vec::new();
+
+    // Process events based on whether id_allocation tuple is provided
+    // Keep original order from event_iter instead of grouping by track/channel
+    let processed_events: Vec<EventUnit<'a, Event>> = event_iter
+        .map(|(&time, event)| {
+            let id = id_allocation
+                .as_mut()
+                .map(|(token_creator, key_extractor, manager)| {
+                    // ID allocation mode: process events with token creator and key extractor
+                    let key = key_extractor(event);
+                    match manager.get_or_new_id(key) {
+                        ObjIdAllocResult::Assigned(id) => Some(id),
+                        ObjIdAllocResult::New(id) => {
+                            let def_token = token_creator(id, key);
+                            late_def_tokens.push(def_token);
+                            Some(id)
+                        }
+                        ObjIdAllocResult::Full => None,
+                    }
+                })
+                .flatten();
```

Comment:
It can be written as below if `get_or_new_id` returns `Option<ObjId>`:
```suggestion
            let id = id_allocation
                .as_mut()
                .and_then(|(token_creator, key_extractor, manager)| {
                    let key = key_extractor(event);
                    if manager.is_assigned(key) {
                        manager.get_or_new_id(key)
                    } else if let Some(new) = manager.get_or_new_id(key) {
                        late_def_tokens.push(token_creator(new, key));
                        Some(new)
                    } else {
                        None
                    }
                });
```

Reply:
I don't want to return whether the id has just allocated now, because it leaks the `ObjIdManager`'s responsibility and internal state.